Bug 1855345 - Make `AutoCopyListener` ignore any selection changes caused by Selection API r=karlt,smaug

Even though there is a reasonable [usage] like bug 1567160, we should
ensure that `AutoCopyListener` never update clipboard when the selection
change is caused by a Selection API call.

Differential Revision: https://phabricator.services.mozilla.com/D191585

This is a security related fix in Firefox 120, but important enough to
include sooner on OpenIndiana.

diff --git a/layout/generic/nsFrameSelection.cpp b/layout/generic/nsFrameSelection.cpp
--- a/layout/generic/nsFrameSelection.cpp
+++ b/layout/generic/nsFrameSelection.cpp
@@ -3323,16 +3323,23 @@ int16_t AutoCopyListener::sClipboardID =
  */
 
 // static
 void AutoCopyListener::OnSelectionChange(Document* aDocument,
                                          Selection& aSelection,
                                          int16_t aReason) {
   MOZ_ASSERT(IsValidClipboardID(sClipboardID));
 
+  // For now, we should prevent any updates caused by a call of Selection API.
+  // We should allow this in some cases later, though. See the valid usage in
+  // bug 1567160.
+  if (aReason & nsISelectionListener::JS_REASON) {
+    return;
+  }
+
   if (sClipboardID == nsIClipboard::kSelectionCache) {
     // Do nothing if this isn't in the active window and,
     // in the case of Web content, in the frontmost tab.
     if (!aDocument || !IsInActiveTab(aDocument)) {
       return;
     }
   }
 
